File: code_extractor.py
import ast
import re

def extract_relevant_snippets(content, language):
    if language.lower() in ['javascript', 'typescript']:
        return extract_js_snippets(content)
    elif language.lower() == 'python':
        return extract_python_snippets(content)
    elif language.lower() == 'css':
        return extract_css_snippets(content)
    else:
        return extract_generic_snippets(content)

def extract_js_snippets(content):
    snippets = []
    # Pattern to match function and class declarations, including async functions and arrow functions
    pattern = r'((?:export\s+)?(?:async\s+)?(?:function|class|const|let|var)\s+\w+\s*(?:\([^)]*\))?\s*{?(?:\s*=>)?(?:\s*{)?(?:\s*\n(?:(?!\n\s*\n).)*)?)'
    matches = re.finditer(pattern, content, re.MULTILINE | re.DOTALL)
    for match in matches:
        snippet = match.group(1).strip()
        # If the snippet is a function or class, include a bit more context
        if re.match(r'(function|class|=>)', snippet):
            # Include up to 5 non-empty lines after the declaration
            lines = snippet.split('\n')
            additional_lines = [line for line in content[match.end():].split('\n') if line.strip()][:5]
            snippet = '\n'.join(lines + additional_lines)
        snippets.append(snippet)
    return snippets

def extract_python_snippets(content):
    snippets = []
    try:
        tree = ast.parse(content)
        for node in ast.walk(tree):
            if isinstance(node, (ast.FunctionDef, ast.ClassDef, ast.AsyncFunctionDef)):
                snippets.append(f"def {node.name}({ast.unparse(node.args)}):" if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)) else f"class {node.name}:")
    except SyntaxError:
        # If parsing fails, fall back to regex
        pattern = r'((?:async\s+)?(?:def|class)\s+\w+\s*(?:\([^)]*\))?:)'
        snippets = re.findall(pattern, content)
    return snippets

def extract_css_snippets(content):
    # Extract CSS rule selectors
    pattern = r'([^{]+)\s*{'
    return re.findall(pattern, content)

def extract_generic_snippets(content):
    # For other languages, extract lines that look like declarations
    pattern = r'^[^=;{}()]+[{(]'
    return re.findall(pattern, content, re.MULTILINE)

File: docgen/src/lib/codeExtractor.js
// Import required modules for regular expressions (no need for ast in JS)
function extractRelevantSnippets(content, language) {
    switch (language.toLowerCase()) {
      case 'javascript':
      case 'typescript':
        return extractJsSnippets(content);
      case 'python':
        return extractPythonSnippets(content);
      case 'css':
        return extractCssSnippets(content);
      default:
        return extractGenericSnippets(content);
    }
  }
  
  function extractJsSnippets(content) {
    const snippets = [];
    // Pattern to match function and class declarations, including async functions and arrow functions
    const pattern = /((?:export\s+)?(?:async\s+)?(?:function|class|const|let|var)\s+\w+\s*(?:\([^)]*\))?\s*{?(?:\s*=>)?(?:\s*{)?(?:\s*\n(?:(?!\n\s*\n).)*)?)/g;
    let match;
    while ((match = pattern.exec(content)) !== null) {
      let snippet = match[1].trim();
      // If the snippet is a function or class, include a bit more context
      if (/function|class|=>/.test(snippet)) {
        // Include up to 5 non-empty lines after the declaration
        const lines = snippet.split('\n');
        const additionalLines = content.slice(match.index + match[0].length).split('\n').filter(line => line.trim()).slice(0, 5);
        snippet = [...lines, ...additionalLines].join('\n');
      }
      snippets.push(snippet);
    }
    return snippets;
  }
  
  function extractPythonSnippets(content) {
    const snippets = [];
    // Fall back to regex for Python parsing (since JS has no `ast`)
    const pattern = /((?:async\s+)?(?:def|class)\s+\w+\s*(?:\([^)]*\))?:)/g;
    let match;
    while ((match = pattern.exec(content)) !== null) {
      snippets.push(match[1].trim());
    }
    return snippets;
  }
  
  function extractCssSnippets(content) {
    // Extract CSS rule selectors
    const pattern = /([^{]+)\s*{/g;
    const snippets = [];
    let match;
    while ((match = pattern.exec(content)) !== null) {
      snippets.push(match[1].trim());
    }
    return snippets;
  }
  
  function extractGenericSnippets(content) {
    // For other languages, extract lines that look like declarations
    const pattern = /^[^=;{}()]+[{(]/gm;
    const snippets = [];
    let match;
    while ((match = pattern.exec(content)) !== null) {
      snippets.push(match[0].trim());
    }
    return snippets;
  }
  
  // Example usage:
  // const content = "Some source code here...";
  // const snippets = extractRelevantSnippets(content, 'javascript');
  // console.log(snippets);
  